### PR TITLE
Fix default icons still being present in the executable

### LINKED ISSUE #870: Do not have a default-icon in the bootloader, add it when freezing
**Original date**: 2014/03/07
**Original reporter**: _AdamS_

In my .spec file, I specify my own icon file using the following syntax:

```
exe = EXE(...
          ...
          icon='myicon.ico')
```

My application builds successfully and displays my own icon. However, when the exe is analysed using a tool such as PE Explorer([http://www.heaventools.com/overview.htm]), Pyinstaller seems to have added two of its own icons to the icon list (see attached for the icons in question).

I guess these icons are added so that they are used as the default in case the user-supplied icon is missing or corrupted. However, a better way to handle this would be to **not** add the Pyinstaller icons to the exe if the user supplies their own.

### PR DESCRIPTION
**Pull request by hdf**
Fixed default icons still being present in the executable after setting a custom icon.
Added --icon=NONE functionality, so that the OS default icon for the file type will be used.

Closes #870 and #2995 (possibly)

### CODE REVIEW

**Comment by Legorooj** [comment]
Can you rebase these commits to fit with our https://pyinstaller.readthedocs.io/en/stable/development/commit-messages.html.

**Comment by hdf** [comment]
Is this acceptable?

**Comment by Legorooj** [comment]
@hdf can you please adhere to the commit link above? All commits must conform.

**Comment by hdf** [comment]
What do you mean? What is wrong with it?

**Comment by Legorooj** [comment]
You message is `Closes 870 ...` whereas that doesn't:
* Use the proper commit prefixes
* Describe what you've done

An example of a commit message which fits that would be:

`Utils: Change the win32 icon handler to delete the icons before assigning a new one`

**Comment by hdf** [comment]
Than is it OK now?

**Comment by Legorooj** [comment]
Yes. Only one thing left: https://pyinstaller.readthedocs.io/en/stable/development/changelog-entries.html

**Comment by hdf** [comment]
I only figured out what you meant, after the push. That "call" word threw me off, I was looking for a function to call with the parameters `add changelog entry`. :) If you said "Just name the changelog ...", I would have added a commit, rather than updating the old one. Oh well, this is fine too, right? (The build fails have nothing to do with my changes.)

**Comment by Legorooj** [comment]
@hdf that build fail is intermittent, and not related to PyInstaller. It's caused by glibc, see #4666 for details. @bjones1 this looks good.

File: PyInstaller/utils/win32/icon.py [hunk 1]
```diff
@@ -145,20 +145,23 @@ def CopyIcons_FromIco(dstpath, srcpath, id=1):
...
-    for i, f in enumerate(icons):
-        data = f.grp_icon_dir()
-        data = data + f.grp_icondir_entries(iconid)
-        win32api.UpdateResource(hdst, RT_GROUP_ICON, i, data)
-        logger.info("Writing RT_GROUP_ICON %d resource with %d bytes", i, len(data))
-        for data in f.images:
-            win32api.UpdateResource(hdst, RT_ICON, iconid, data)
-            logger.info("Writing RT_ICON %d resource with %d bytes", iconid, len(data))
-            iconid = iconid + 1
+    hdst = win32api.BeginUpdateResource(dstpath, 1)
```
Comment (bjones1) [inline PyInstaller/utils/win32/icon.py]: Per https://pywin32-ctypes.readthedocs.io/en/stable/api/win32ctypes.pywin32.win32api.html#win32ctypes.pywin32.win32api.BeginUpdateResource, the second parameter should be True, not 1. Add a comment along the lines of "Remove any existing resources from the file." Is this OK -- are there any resources embedded in the file we don't want to delete?

File: news/4700.feature.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+Change the win32 icon handler to delete the icons before assigning a new one.
+Also, using --icon=NOICON or --icon=NONE allows you to use the OS default icon.
```
Comment (bjones1) [inline news/4700.feature.rst]: Pick one or the other, but not both. Also, the manual and command-line help text needs to be updated to reflect this new option.

File: PyInstaller/utils/win32/icon.py (hunk 1 above)
Comment (hdf) [reply PyInstaller/utils/win32/icon.py]: It was 0 before, so I figured, a 1 would be acceptable now, but OK.
Icon handling is the first thing we do with the resources, so deleting everything at this point should be fine, I checked.

Comment (Legorooj) [reply PyInstaller/utils/win32/icon.py]: @bjones1 no it should be fine.

**Comment by hdf** [comment]
@bjones1 Is it OK now?

**Comment by bjones1** [comment]
Have you tested this code?

**Comment by hdf** [comment]
Of course. It works for me wonderfully. Already using it.

**Comment by bjones1** [comment]
Sadly, I must stop working on this for the foreseeable future. All my classes must now be taught online, and trying to make that happen will take all my time.

I did a bit of testing earlier, in which this code actually cased Windows to delete the executable, but I don't have time to debug.

**Comment by Legorooj** [comment]
@bjones1 I'll have a look as well if I've a moment.

**Comment by hdf** [comment]
Sorry, I only tested the icon reset, not the --icon=NONE functionality. It is fixed now.

Btw., on the way I found a new bug:
`AttributeError: module 'win32ctypes.pywin32.win32api' has no attribute 'error'`
It is caused by this line:
https://github.com/pyinstaller/pyinstaller/blob/60d00dccb01b2aab5fdd9885932729ce774551b5/PyInstaller/utils/win32/icon.py#L234

**Comment by htgoebel** [comment]
@hdf Thansk for working on this long-standing issue.
Why not take a different approach:
- change the bootloader build to not add a default icon to the bootloader. This saves the need for code for removing any existing icon.
- *Internally* change the default of `self.icon` to the respective icon from `bootloader/images/*.ico` in([PyInstaller/building/api.py, line 350](https://github.com/pyinstaller/pyinstaller/blob/v3.6/PyInstaller/building/api.py#L350))
This should solve #870. 

> Added --icon=NONE functionality, so that the OS default icon for the file type will be used.

This is a different issue, isn't it? So I suggest opening a new issue for this. (Issue, since I strongly doubt, adding no icon the the *executable* will make the executable to get some file-types's default icon.)

> AttributeError: module 'win32ctypes.pywin32.win32api' has no attribute 'error'

If this persist after the change I suggested, I suggest opening a new pull-request, do, since this should occur independent of this very change.

**Comment by hdf** [comment]
@htgoebel Because, modifying the bootloader is harder, and more complicated. And resetting the resource is easy, it's just changing a 0 to a True.

> This is a different issue, isn't it? So I suggest opening a new issue for this. (Issue, since I strongly doubt, adding no icon the the executable will make the executable to get some file-types's default icon.)

Since resetting is so easy, having no icon is easy too, and it is related.
The default OS icon isn't baked into the executable, the OS assigns it, when no icon is found in the executable.

> If this persist after the change I suggested, I suggest opening a new pull-request, do, since this should occur independent of this very change.

Technically it is unrelated, but it is just 2 lines, and this is an ancient problem as well, and this is the only place it comes up, with the icon handling, so not totally unrelated. Is it being its own commit not enough? Does it have to be it's own push request?

**Comment by hdf** [comment]
This linter is pissing me off. The build fails because `module level import not at top of file`. Except, I did not introduce that problem, did I? So why was it not throwing an error before? The whole code is full of linting violations, that it ignores, but if I modify something near one, suddenly it notices them. So, how do I make the linter ignore my stuff? Or shall I go fix every linting error in the file?

**Comment by Legorooj** [comment]
@hdf you'll need to fix every violation in the file. (I feel you; it's happened to me as well - in a file I didn't change)

**Comment by htgoebel** [comment]
> Because, modifying the bootloader is harder, and more complicated. 

Yet, it is a bit harder, but not complicated. Just do not add the icon resource when building. And it is a much cleaner approach, also saving space since the bootloader will become smaller.

I added some inline comments, explaining why your approach does not convince me.

> This linter is pissing me off. The build fails because module level import not at top of file. Except, I did not introduce that problem, did I? So why was it not throwing an error before?

The linter works on changed lines only, not on the whole file. This is why it triggered on your change, even if the error was there before. Just ignore it, thiw will net affect whether we merge. (If the linter would consider the whole file, we would have too many errors. And we have the policy to only accept functional changes. Fixing lint-errors would be non-functional. But you can ignore this error, we will merge even with it.)

Beside this:
- Please add a changelog entry for 870 and 2995, explaining what bug has been fixed (Currently there is only a changelog entry for a "feature".
- We need at least a simple test-case for `--icon=NONE`, see tests/functional/test_basic.py:test_set_icon
- `icon=NONE` needs to be documented in docs/usage.rst (mind the OSX case!)
- Does `icon=NONE` make sense on OS X?

As you can see, for feature, there is much more to do than for a bugfix, thus a different pull-request would be worth IMHO.

**Comment by Legorooj** [comment]
@htgoebel about non-functional changes; there'll be a few in my py2cleanup PR (#4729), because a *lot* of the code base needs to be cleaned thoroughly (So much was inefficient because of py2).

**Comment by hdf** [comment]
@htgoebel If we build the bootloader without an icon by default, than won't the default PyInstaller icon never be used? (Unless the user specifies it, with path, manually.) This does not seem like a good approach to me. (Although the necessary change might simply be to comment out line 31 in `bootloader/windows/run.rc` and `runw.rc`.)

Where are these mentioned inline comments? I can't see them.

I don't have access to a Mac, so all my changes are for Windows only. (I suppose I should mention that in the docs.)

How should I test to see if the built executable has no icon? (Maybe `len(win32api.EnumResourceNames(win32api.LoadLibraryEx("run.exe", 0, 2), 3)) == 0`?)
